web/webapi.py

```diff
@@ -418,52 +412,178 @@ def header(hdr, value, unique=False):
     ctx.headers.append((hdr, value))
 
 
-def rawinput(method=None):
-    """Returns storage object with GET or POST arguments."""
-    method = method or "both"
+class MultipartPartWrapper:
+    """
+    Ensures calling `.value` on a multipart file object returns the raw value
+    in the same way that calling `.value` on FieldStorage multipart objects
+    did.
+
+    With MultipartPart, `.value` decodes the data, which will fail with raw
+    byte sequences, such as images.
+
+    Additionally, MultipartPart lacks a setter, so it wasn't possible to rebind
+    MultipartPart's `.value` method with the `.raw` method. This passes
+    through all attribute/method access to the original MultipartPart object,
+    save for `.value`, which it intercepts and for which it returns the `.raw`
+    method.
+
+    With this, the following code should behave as it did with
+    `cgi.FieldStorage`:
+
+    class ImageUpload:
+        def POST(self, path):
+            i = web.input(file={})
+            name, filename, value = i.file.name, i.file.filename, i.file.value
+    """
+
+    def __init__(self, part: MultipartPart) -> None:
+        self._part = part
+
+    def __getattr__(self, name: str) -> Any:
+        """
+        Return raw data when accessing the list property, or otherwise pass
+        the request through.
+        """
+        if name == "value":
+            return self._part.raw
+
+        return getattr(self._part, name)
+
+
+def parse_multipart_data(
+    stream: BytesIO, boundary: str, content_length: int, **kwargs
+) -> tuple[MutableMapping, MutableMapping]:
+    """
+    For compatibility with cgi.FieldStorage, if all uploads use the
+    same name, then put them under in list under a single key. E.g.:
+    {'file': [<file_object_1>, <file_object_2>]}
+    But if the files use different keys, just put each object under
+    its own key. E.g.:
+    {'file_1': <file_object_1>, 'file_2': <file_object_2>}
+    """
+    files = defaultdict(list)
+    forms = defaultdict(list)
+
+    for part in MultipartParser(stream, boundary, content_length, **kwargs):
+        if part.filename or not part.is_buffered():
+            files[part.name].append(part)
+        else:
+            forms[part.name].append(part.value)
+
+    # Flatten lists with only one item for cgi.FieldStorage compatibility.
```

Comment:
If someone is using web.py still (which is itself at this stage somewhat of a legacy tool) I think it's not unreasonable to assume those projects relying on it may not be super active and if we can avoid requiring anyone to update their code (including the Open Library project) I can see why this would be an advantage. It seems useful to centralize all of the changes in one repo rather than require every project to change its code downstream.

Reply:
I guess I should have said **attempt** to preserve **partial** compatibility with the internal implementation (not external API). `MultiPartWrapper` is an invention no one has ever seen before. 

If someone is expecting the internal implementation to be the same, they're out of luck anyway because `MultiPartWrapper`  != `cgi.FieldStorage` (and the latter is gone, so they're going to have to modify their code anyway).